### DISCUSSION TITLE
Type resolution, type IDs, checkcast/instanceof, native types

### DISCUSSION THREAD

**Discussion by dmlloyd** [qbicc/qbicc#430] [General research]
I'm in the process of refactoring and simplifying the `DescriptorTypeResolver` system.  The general thrust of the work is to resolve descriptors to `ObjectType`s rather than `ReferenceType`s for most cases, and propagating the resultant simplifications throughout the code base, including through `CheckCast` and `InstanceOf` node types.  This work along with @dgrove-oss's work on `CheckCast` and @DanHeidinga's work on `InstanceOf` has reached a point of convergence that it is now time to resolve.  This is a summary of previous discussion mixed with potential improvements that, if agreed, ultimately should make its way back into the design documentation.

Ideally, both of these node types would accept a value input and a type input.  The type of the `CheckCast` node would be the value's type narrowed by the static type of the type input.  Recently we've discussed making the type argument more dynamic, and @dgrove-oss has done so for the `CheckCast` node.  However the implementation had to depart from the ideal for a few reasons, the most relevant being: there is more than one failure type for cast failures depending on the situation, and: for array storage checks, we do not have a good way (or really *any* way) to represent reference array type IDs at run time.  It is this latter problem that must now be solved because it pervades not only `CheckCast` but also `InstanceOf`.

The current `CheckCast` implementation relies on having different semantics for array store casts versus common casts.  However, it does not account for cases like casting an array value to another array type, which have similar requirements in terms of validating the type of the array in question.  Specifically the problem revolves around comparing the effective run-time types of one or two reference array types.

Reference array types may be the source or target of both `CheckCast` or `InstanceOf` operations.  There are cases where the type in question may be static or may be dynamic.  We can only support so many functional permutations before things get out of hand.

I propose that these two operation nodes be refined to account *generally* for the possibility of checks or casts to array types that may either be statically known or known only at run time.  Since reference array types do not have an integer type ID, I propose that these node types be redefined with parameters of the following type:

| Type | Name | Value Type | Description |
| --- | --- | --- | --- |
| `Value` | `input` | `ReferenceType` | The input value which must be a reference. |
| `Value` | `toType` | `TypeType` | The type to cast to (may be a `TypeLiteral`).  The upper bound of the type must be `ClassObjectType`, `PrimitiveArrayObjectType`, or `InterfaceObjectType`. |
| `Value` | `toDimensions` | `IntegerType` | The number of array dimensions (may be an `IntegerLiteral`, must be zero if the argument type is not a reference array). |
| `CastType` | `castType` | *n/a* | The kind of exception to throw on cast failure (`CheckCast` only). |

No type IDs for reference array types would ever be handled; instead, such cases would be expanded to the leaf type for `toType` and the number of dimensions of the reference array type for `toDimensions`.

Native types would formally no longer use `CheckCast` or `InstanceOf` nodes in *any* capacity; any use of `checkcast` or `instanceof` opcodes would be handled before the member resolver and would be transformed to word casts and static type checks.  Thus for the node types we can safely require (with appropriate enforcement) the types of the parameter values to conform as specified above.

Since both node types use the same prototype, lowering `CheckCast` could be consistently simple:

```
// current = "startBlock"
If(IsEq(input, null), okay, next)
Begin(next)
If(InstanceOf(input, toType, toDimensions), okay, fail)
Begin(fail)
Invoke(failMethod)
Unreachable()
Begin(okay)
// then:
outputType := input.type.narrow(toType.array(toDimensions).upperBound)
output := φ(startBlock → null of outputType, next → BitCast(input, outputType.notNull));
```

This can probably be done inline since one or both of the `If` may be trivially optimizable (hoisting to an inlinable method is also a possibility but doing so could make it somewhat more likely to spoil optimization opportunities).

Therefore, the complex logic would generally sit within the handling of `InstanceOf`.  Unfortunately the number of cases is not small, however it should at least be possible to get most if not all of the complexity into one single location.

Here are the set of potentially interesting cases - and there are plenty of them.  The term "out-of-bounds" refers to whether a given literal type ID is known to be outside of the possible type ID range for a given dynamic type ID.

The `input` value may be:
* `null` (literal) (always `False`)
* A non-reference-array object literal (future but soon)
* A reference-array object literal (future but soon)
* A definitely non-reference-array object reference
* A definitely reference-array object reference whose leaf type is `Object` and whose run-time dimensions are `m` (i.e. a supertype of all reference-array objects whose dimensions are ≥ `m`)
* A definitely reference-array object reference whose leaf type is `Serializable`, or `Cloneable`, and whose run-time dimensions are `m` (i.e. a supertype of all reference-array objects whose dimensions are strictly > `m`)
* A definitely reference-array object reference whose run-time dimensions are `m` and whose leaf type is not `Object`, `Serializable`, or `Cloneable`
* A reference whose physical upper bound is `Object` and whose interface bounds are any subset of { `Cloneable`, `Serializable` } including the empty set (i.e. a type which may or may not be a reference-array object)

The `toType` value may be:
* The `Object` `TypeLiteral` (maps to `IsNe(input, null)` always)
* A `TypeLiteral` whose set of subtypes is strictly disjoint with respect to the set of subtypes of `input`'s type (always `False`; may be difficult to deduce)
* A `ClassObjectType` or `PrimitiveArrayObjectType` `TypeLiteral` (both cases should be similar)
* An `InterfaceObjectType` `TypeLiteral`
* A dynamic type ID

The `toDimensions` value may be:
* Zero (literal)
* Literal `q` where `q > 0`
* A dynamic value `q` which may or may not satisfy `q ≥ 0` (run-time checked, potentially)
* A dynamic value `q` with a statically known minimum bound of `r`, where `r ≥ 0` (future, see #371)

These two constructs should be sufficient to support not only the `instanceof` and `checkcast` opcodes and the `Class.cast` operation, but also `Class.isInstance` and `Class.isAssignableFrom` for all valid reference types, unifying all similar cases.  However, there are still cases that would still necessarily have to be handled with slow-path methods.

For example, the case where a value which may be an array is checked against a type which may be an array type is very different than if the type values are all known statically.  The `InstanceOf` check would only succeed after several checks for combinations of the above cases, some of which may or may not be able to be statically deduced.

Implementing an `InstanceOf` that is aware of every combination of possibly interesting inputs may be too complex to be feasible.  The current implementation is already fairly complex, and it explicitly leaves out several cases and is implemented in terms of a static type ID.  It may be more worthwhile to define the bulk of the `InstanceOf` logic in a method or series of methods, implemented pessimistically for the fully dynamic case, but strategically so, with a view towards allowing the inliner and simple optimizer to eliminate constant conditions on a per-callsite basis.  This would avoid the previously discussed potential situation where we are reimplementing numerous constant condition checks that are already handled by other stages of compilation.

**Comment by dgrove-oss** [comment]
I would suggest changing `dimensions` to `toDimensions` because it has the same logical role as `toType` in the operation.

**Comment by dgrove-oss** [comment]
It's Friday afternoon, and this code doesn't compile because I am using imaginary intrinsics, but here's what VMHelpers (out-of-line code) might look like...
```java
    public static int full_instanceof(Object instance, Class<?> castClass) {
        if (instance == null) {
            return 0;
        }
        type_id toTypeId = ObjectModel.get_type_id_from_class(castClass);
        int toDim = ObjectModel.get_dim_from_class(castClass);
        return isAssignableTo(instance, toTypeId, toDim) ? 1 : 0;
    }

    public static void arrayStoreCheck(Object value, type_id toTypeId, int toDimensions) {
        if (value == null || isAssignableTo(value, toTypeId, toDimensions - 1)) {
            return;
        }
        raiseArrayStoreException();
    }

    public static void checkcast_class (Object value, Class<?> cls) {
        checkcast_typeId(value, ObjectModel.get_type_id_from_class(cls), ObjectModel.get_dim_from_class(castClass));
    }

    public static void checkcast_typeId(Object value, type_id toTypeId, int toDimensions) {
        if (value == null || isAssignableTo(value, toTypeId, toDimensions)) {
            return;
        }
        throw new ClassCastException();
    }

    // Invariant: value is not null
    private static boolean isAssignableTo(Object value, type_id toTypeId, int toDimensions) {
        if (toDimensions == 0) {
            return isAssignableToLeaf(ObjectModel.type_id_of(value), toTypeId)
        } else if (ObjectModel.isReferenceArray(ObjectModel.type_id_of(value))) {
            int valueDims = ObjectModel.dimensions_of(value);
            if (valueDims == toDimensions) {
                type_id valueElemTypeId = ObjectModel.element_type_id_of(value);
                return isAssignableToLeaf(valueElemTypeId, toTypeId);
            } else if (valueDims > toDimensions) {
                return ObjectModel.is_java_lang_object(toTypeId);
            }
        }
        return false;
    }

    private static final boolean isAssignableToLeaf(type_id valueTypeId, type_id toTypeId) {
        if (ObjectModel.isClass(toTypeId)) {
            type_id maxTypeId = ObjectModel.max_subclass_type_id_of(toTypeId);
            return toTypeId.intValue() <= valueTypeId.intValue() && valueTypeId.intValue() <= maxTypeId.intValue());
        } else if (ObjectModel.isInterface(toTypeId)) {
            return ObjectModel.doesImplement(valueTypeId, toTypeId);
        } else {
            // PrimitiveObjectArray
            return valueTypeId == toTypeId;
        }
    }
```

**Comment by dgrove-oss** [comment]
Reached a reasonably complete implementation of this design in #434.
